rasa/core/trackers.py

```diff
@@ -420,11 +432,61 @@ def undo_till_previous(event_type, done_events):
                 # the `action_listen` action right before it).
                 undo_till_previous(UserUttered, applied_events)
                 undo_till_previous(ActionExecuted, applied_events)
+            elif (
+                isinstance(event, ActionExecuted)
+                and event.action_name in form_names
+                and not self._first_loop_execution_or_previous_rejection(
+                    event.action_name, applied_events
+                )
+            ):
+                self.undo_till_previous_loop_execution(
+                    event.action_name, applied_events
+                )
             else:
                 applied_events.append(event)
 
         return applied_events
 
+    @staticmethod
+    def _first_loop_execution_or_previous_rejection(
+        loop_action_name: Text, applied_events: List[Event]
+    ) -> bool:
+        for event in reversed(applied_events):
+            # Form was rejected previously.
+            if (
+                isinstance(event, ActionExecutionRejected)
+                and event.action_name == loop_action_name
+            ):
+                return True
+
+            # Form might have run before but was deactivated and is now running again.
+            if isinstance(event, Form) and event.name is None:
+                return True
+
+            # Form ran before.
+            if (
+                isinstance(event, ActionExecuted)
+                and event.action_name == loop_action_name
+            ):
+                return False
+
+        return True
+
+    @staticmethod
+    def undo_till_previous_loop_execution(
```

Comment:
Good point. I'll check what the best practices are and make it consistent then.

Reply:
I'll make both private class methods since there is no need to protect this function from the rest of the world and they are no closures (https://stackabuse.com/python-nested-functions)

Reply:
```suggestion
    def _undo_till_previous_loop_execution(
```
I think this should be private